### PR TITLE
Move driver setup button under driver selection #976

### PR DESCRIPTION
**Pull request by ann0see**
Solves #973

~~The tooltip of this still tells you about the buffer settings. I'm not sure how to change that.~~ Done

### CODE REVIEW

**Comment by pljones** [comment]
How to find out:
```
peter@fs-peter:~/git/Jamulus-wip$ grep -nr butDriverSetup src
src/clientsettingsdlgbase.ui:245:          <widget class="QPushButton" name="butDriverSetup">
src/clientsettingsdlgbase.ui:677:  <tabstop>butDriverSetup</tabstop>
src/clientsettingsdlg.cpp:185:    butDriverSetup->setWhatsThis ( strSndCrdBufDelay );
src/clientsettingsdlg.cpp:186:    butDriverSetup->setAccessibleName ( tr ( "ASIO setup push button" ) );
src/clientsettingsdlg.cpp:187:    butDriverSetup->setToolTip ( strSndCrdBufDelayTT );
src/clientsettingsdlg.cpp:281:    butDriverSetup->setText ( tr ( "ASIO Setup" ) );
src/clientsettingsdlg.cpp:284:    butDriverSetup->hide();
src/clientsettingsdlg.cpp:418:    QObject::connect ( butDriverSetup, &QPushButton::clicked,
```
So 185, 186, 187.

**Comment by ann0see** [comment]
Ok. Thanks! Now the question is: What should the text be?

My first thought:

```

    // Driver setup button
    QString strSndCardDriverSetup = "<b>" tr ( "Sound card driver settings" ) + "</b>" +
        tr ( "The sound card driver settings button opens the driver settings pannel of " +
        "your sound card. Some drivers allow you to change buffer settings, others like the " +
        "ASIO4ALL let you choose input or outputs of your device(s). To show the inputs or outputs " +
        "in ASIO4ALL, click on the tool icon (= advanced view) on the bottom right, " +
        "click the plus icon left to the sound card and afterwards enable the correct input and output. " +
        "By hovering over an input output under a soundcard, you can see if it is an input or output. " +
        "Make sure to only enable devices which support a sample rate of " ) + SYSTEM_SAMPLE_RATE_HZ +
        tr ( "Hz. To check this in ASIO4ALL, hover over the device or input/output and check the " +
        "supported sample rate. New devices will not automatically be added to ASIO4ALL. Therefore, " +
        "please restart " ) + APP_NAME + tr ( " if you connect new devices." );

    QString strSndCardDriverSetupTT = tr ( "Opens the driver settings. Note: " + APP_NAME + " currently " +
        "only supports devices supporting a sample rate of " + SYSTEM_SAMPLE_RATE_HZ "Hz. " +
        "You will not be able to select a driver/device which doesn't. " +
        "If you use ASIO4ALL, you might need to choose the correct input output devices which also support " +
        "this sample rate. Check this by hovering over a device. Selecting devices " +
        "in ASIO4ALL can be done by clicking here, afterwards in ASIO4ALL clicking on the tool icon "
        "(advanced view), then on the plus icon left to the sound card. Now select the correct input or output "
        "devices. For more help see the Website." ) + TOOLTIP_COM_END_TEXT;
```

**Comment by pljones** [comment]
I would not add anything specific to any driver.  Reference the web site for more guidance on setting up your sound card but nothing deeper than that.

**Comment by ann0see** [comment]
Ok. So I'd be happy about a suggestion from you and @gilgongo

**Comment by ann0see** [comment]
What about:

```

    // Driver setup button
    QString strSndCardDriverSetup = "<b>" tr ( "Sound card driver settings" ) + "</b>" +
        tr ( "The sound card driver settings button opens the driver settings pannel of " +
        "your sound card. Some drivers allow you to change buffer settings, others like the " +
        "ASIO4ALL let you choose input or outputs of your device(s)." );

    QString strSndCardDriverSetupTT = tr ( "Opens the driver settings. Note: " + APP_NAME + " currently " +
        "only supports devices supporting a sample rate of " + SYSTEM_SAMPLE_RATE_HZ "Hz. " +
        "You will not be able to select a driver/device which doesn't. " +
        "For more help see the Website." ) + TOOLTIP_COM_END_TEXT;
```

**Comment by pljones** [comment]
I've not thought of anything better myself.  I'd still be inclined to say less, though, on `strSndCardDriverSetup` -- maybe completely lose the last sentence?  It should just say it gets you to the sound card settings - I feel saying more is making it sound like it's somehow something Jamulus controls.

**Comment by ann0see** [comment]
Sure, It can be removed. I might need to add my explanation to this:

My intention behind adding ASIO4ALL was to help new users (who probably use ASIO4ALL and will therefore click this button and open this help text), should at least receive some help here. That's why I initially wrote about this information. We might need to think a bit more of ASIO4ALL first user problems. That's probably a topic for another issue.

**Comment by pljones** [comment]
Yep, I understand the intent - but for non-ASIO4ALL users, it could be off-putting / misleading.  I think things like this are better handled in "first time user" hand-holding.

**Comment by ann0see** [comment]
>  I think things like this are better handled in "first time user" hand-holding.

... which doesn't exist yet ;-). But of course, this is another issue.

Ok. So maybe your suggestion should be applied then? Any other thoughts?

**Comment by gene96817** [comment]
I appreciate the level of attention and effort for addressing issues with AISO4ALL.

MacOS users need to be assured they do not have similar or related problems.

With all the discussion about AISO4ALL, how should Windows users think about manufacture specific AISO drivers. The current documentation suggest they use the AISO4ALL driver. I have seen the Dell AISO driver perform better. Could a user see more than three ASIO drivers installed in their machine?

**Comment by ann0see** [comment]
> I have seen the Dell AISO driver perform better.

Just a short question: is it a generic driver too?


> Could a user see more than three ASIO drivers installed in their machine?

Yes. I don't know why they shouldn't.

**Comment by gene96817** [comment]
The Dell ASIO appears to only know about the built-in audio circuit.

**Comment by ann0see** [comment]
Ok. Can you link it? I'd like to test it on my HP PC. Maybe it's compatible?

**Comment by hoffie** [comment]
> Ok. Can you link it? I'd like to test it on my HP PC. Maybe it's compatible?

Just found this: https://www.dell.com/support/driver/de/de/debsdt1/driversdetails?driverid=7776f

There are also attempts at getting similar things running by editing INF files:
https://www.tenforums.com/sound-audio/150202-realtek-asio-driver-installs-but-fails-operate.html

I have no experience with this. Maybe it'll work better. Doesn't sound easier to set up than ASIO4ALL at the first glance though.

**Comment by pljones** [comment]
> Yes. I don't know why they shouldn't.

Yep - I've got six on one of my two machines (not the one I actually use for jamming, either!).

**Comment by ann0see** [comment]
Note to self: syntax error.

**Comment by ann0see** [comment]
Ok. Ready for review. 
I'd especially like to have a feedback on the description here from @gilgongo

@pljones 
Should this be part of the next release or should we still wait? I'm a bit afraid that existing users might be confused with the change here. New users in contrary would probably find and understand the new position faster.

**Comment by pljones** [comment]
I think _most_ users won't notice the difference, to be honest.  If their settings are working, they won't be looking at the ASIO and audio buffer bits.  I think it's pretty safe to merge.

**Comment by pljones** [comment]
Can you confirm that you have tested this on Linux with JACK and that the hover text and What's This still make sense there?

**Comment by ann0see** [comment]
This doesn't even show up on Linux. I will re-test it.

**Comment by ann0see** [comment]
JamulusLinuxSettings

**Comment by gilgongo** [comment]
Would it be better to use "this" to refer to the element the user is looking at? Maybe also remove the redundant "panel". Shorter at least:

"This opens the driver settings of your sound card. Some drivers allow you to change buffer settings, others like the ASIO4ALL let you choose input or outputs of your device(s)"

**Comment by ann0see** [comment]
@gilgongo now I use your suggestion.

File: src/clientsettingsdlgbase.ui [hunk 1]
```diff
@@ -674,7 +675,6 @@
   <tabstop>rbtBufferDelayPreferred</tabstop>
   <tabstop>rbtBufferDelayDefault</tabstop>
   <tabstop>rbtBufferDelaySafe</tabstop>
-  <tabstop>butDriverSetup</tabstop>
```
Comment (ann0see) [inline src/clientsettingsdlgbase.ui:677]: @mulyaj I think we need to re-add this here. So you'd add 
```xml
 <tabstop>cbxSoundcard</tabstop>
 <tabstop>butDriverSetup</tabstop>
```

right under `<tabstops>`

**Review by ann0see — COMMENTED** [review]
@pljones @mulyaj added a few suggestions here.

File: src/clientsettingsdlgbase.ui (hunk 1 above)
Comment (ann0see) [reply src/clientsettingsdlgbase.ui:677]: In fact, it seems to work without this too. Nevertheless, I will do some more tests.

File: src/clientsettingsdlg.cpp [hunk 2]
```diff
@@ -169,12 +169,12 @@ CClientSettingsDlg::CClientSettingsDlg ( CClient*         pNCliP,
     QString strSndCrdBufDelayTT = tr ( "If the buffer delay settings are "
         "disabled, it is prohibited by the audio driver to modify this "
         "setting from within the software. "
-        "On Windows, press the ASIO Setup button to open the "
+        "On Windows, press the ASIO Device Setup button to open the "
```
Comment (mulyaj) [inline src/clientsettingsdlg.cpp:172]: FYI I renamed the button name to _ASIO Device **settings**_.

Comment (ann0see) [reply src/clientsettingsdlg.cpp:172]: Ah. Sorry. I will revert and rename it to the correct name

Comment (mulyaj) [reply src/clientsettingsdlg.cpp:172]: The tooltips seemed to mention settings more often, so I just kept it consistent

Comment (ann0see) [reply src/clientsettingsdlg.cpp]: Ok. Can you please look at the new changes?

Comment (mulyaj) [reply src/clientsettingsdlg.cpp]: looks good

Comment (ann0see) [reply src/clientsettingsdlg.cpp]: Great. Thanks. I will resolve this now.

File: src/clientsettingsdlgbase.ui (hunk 1 above)
Comment (pljones) [reply src/clientsettingsdlgbase.ui:677]: Please keep them.  They help accessibility tools, as far as I'm aware.

Comment (ann0see) [reply src/clientsettingsdlgbase.ui:677]: I'm also quite sure that they should stay. @mulyaj or me can add them (but I'll not do it before tomorrow)

Comment (mulyaj) [reply src/clientsettingsdlgbase.ui:677]: I'll add them back

**Comment by ann0see** [comment]
Ok. It doesn't seem to change anything. But that's ok I think. So from my point of view, this is ready to be merged now.

**Review by gilgongo — APPROVED** [review]

**Review by pljones — APPROVED** [review]

**Pull request merged**